core/src/interest.rs

```diff
@@ -0,0 +1,275 @@
+        Ok(RangeOpen { start, end })
+    }
+
+    /// Report the not after value
+    pub fn not_after(&self) -> Result<u64> {
+        let mut decoder = Decoder::from_bytes(self.0.as_slice());
+        // Skip peer_id
+        decoder.items().next();
+        // Skip start
+        decoder.items().next();
+        // Skip end
+        decoder.items().next();
+
+        decode_u64(&mut decoder)
+    }
+
+    /// Return the interest as a slice of bytes.
+    pub fn as_slice(&self) -> &[u8] {
+        self.0.as_slice()
+    }
+}
+
+fn decode_bytes(decoder: &mut Decoder<Cursor<Vec<u8>>>) -> Result<Vec<u8>> {
+    if let Some(item) = decoder.items().next() {
+        let item = item?;
+        match item {
+            cbor::Cbor::Bytes(data) => Ok(data.0),
+            item => Err(anyhow!("expected cbor bytes, found: {:?}", item)),
+        }
+    } else {
+        Err(anyhow!("expected top level cbor value, found nothing"))
+    }
+}
+fn decode_u64(decoder: &mut Decoder<Cursor<Vec<u8>>>) -> Result<u64> {
+    if let Some(item) = decoder.items().next() {
+        let item = item?;
+        match item {
+            cbor::Cbor::Unsigned(data) => Ok(data.into_u64()),
+            item => Err(anyhow!("expected cbor unsigned integer, found: {:?}", item)),
+        }
+    } else {
+        Err(anyhow!("expected top level cbor value, found nothing"))
+    }
+}
+
+impl Display for Interest {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{}", multibase::encode(Base::Base58Btc, &self.0))
+    }
+}
+impl FromStr for Interest {
+    type Err = anyhow::Error;
+
+    fn from_str(s: &str) -> std::result::Result<Self, Self::Err> {
+        let (_, bytes) = multibase::decode(s)?;
+        // TODO validate these bytes are valid Interest bytes
+        Ok(Self(bytes))
+    }
+}
+
+/// Builder provides an ordered API for constructing an EventId
+#[derive(Debug)]
+pub struct Builder<S: BuilderState> {
+    state: S,
+}
+/// The state of the builder
+pub trait BuilderState {}
+
+/// Initial state of the builder.
+#[derive(Debug)]
+pub struct Init;
+impl BuilderState for Init {}
+
+/// Builder with sort key set
+pub struct WithSortKey {
+    encoder: Encoder<Vec<u8>>,
+}
+impl BuilderState for WithSortKey {}
+
+/// Builder with peer id set
+pub struct WithPeerId {
+    encoder: Encoder<Vec<u8>>,
+}
+impl BuilderState for WithPeerId {}
+
+/// Builder with range key set
+pub struct WithRange {
+    encoder: Encoder<Vec<u8>>,
+}
+impl BuilderState for WithRange {}
+
+/// Builder with not after set
+pub struct WithNotAfter {
+    encoder: Encoder<Vec<u8>>,
+}
+impl BuilderState for WithNotAfter {}
+
+impl Builder<Init> {
+    pub fn with_sort_key(self, sort_key: &str) -> Builder<WithSortKey> {
+        let mut hasher = Sha2_256::default();
+        hasher.update(sort_key.as_bytes());
+        // sha256 is 32 bytes safe to unwrap to [u8; 32]
+        let hash: [u8; 32] = hasher.finalize().try_into().unwrap();
```

Comment:
So later on, this gets converted to a vec, just taking the last 8 items. Should just return the finalize, then take a part of the vec.